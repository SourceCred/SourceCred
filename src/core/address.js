// @flow

import deepEqual from "lodash.isequal";
import stringify from "json-stable-stringify";

export type Address = {|
  +pluginName: string,
  +id: string,
  +type: string,
|};

export interface Addressable {
  +address: Address;
}

export type SansAddress<T: Addressable> = $Exact<$Diff<T, {+address: Address}>>;

export type AddressMapJSON<T: Addressable> = {
  [serializedAddress: string]: SansAddress<T>,
};

/**
 * A data structure for storing addressable objects, keyed by their
 * addresses.
 */
export class AddressMap<T: Addressable> {
  /*
   * Nested structure for fast access.
   *
   * It is an representation invariant that there are no empty objects
   * in this structure (except possibly the top-level object). In
   * particular, if `_data[somePluginName]` is not `undefined`, then it
   * is a non-empty object. This is required so that two `AddressMap`s
   * are logically equal exactly if their `_data` fields are deep-equal.
   *
   * This nested structure is significantly more performant than a
   * simpler version using a flat object keyed by `stringify(address)`.
   * For basic performance tests, see:
   *     https://jsperf.com/address-string-302039074
   */
  _data: {[pluginName: string]: {[type: string]: {[id: string]: T}}};

  /**
   * Create an empty `AddressMap`.
   */
  constructor() {
    this._data = {};
  }

  /**
   * Test whether this map logically equals another map. Two maps are
   * logically equal if they contain the same keys and the values at
   * each key are deep-equal.
   */
  equals(that: AddressMap<T>): boolean {
    return deepEqual(this._data, that._data);
  }

  toJSON(): AddressMapJSON<T> {
    const result = {};
    Object.keys(this._data).forEach((pluginName) => {
      const dataForPluginName = this._data[pluginName];
      Object.keys(dataForPluginName).forEach((type) => {
        const dataForType = dataForPluginName[type];
        Object.keys(dataForType).forEach((id) => {
          const address = {pluginName, id, type};
          const datum = {...dataForType[id]};
          delete datum.address;
          result[toString(address)] = datum;
        });
      });
    });
    return result;
  }

  static fromJSON(json: AddressMapJSON<T>): AddressMap<T> {
    const result: AddressMap<T> = new AddressMap();
    Object.keys(json).forEach((key) => {
      result.add({...json[key], address: JSON.parse(key)});
    });
    return result;
  }

  /**
   * Add the given object to the map, replacing any existing value for
   * the same address.
   *
   * Returns `this` for easy chaining.
   */
  add(t: T): this {
    const {address} = t;
    if (address == null) {
      throw new Error(`address is ${String(t.address)}`);
    }
    let dataForPluginName = this._data[address.pluginName];
    if (dataForPluginName === undefined) {
      this._data[address.pluginName] = dataForPluginName = {};
    }
    let dataForType = dataForPluginName[address.type];
    if (dataForType === undefined) {
      dataForPluginName[address.type] = dataForType = {};
    }
    dataForType[address.id] = t;
    return this;
  }

  /**
   * Get the object at the given address, if it exists, or `undefined`
   * otherwise.
   */
  get(address: Address): T {
    if (address == null) {
      throw new Error(`address is ${String(address)}`);
    }
    const dataForPluginName = this._data[address.pluginName];
    if (dataForPluginName === undefined) {
      return (undefined: any);
    }
    const dataForType = dataForPluginName[address.type];
    if (dataForType === undefined) {
      return (undefined: any);
    }
    return dataForType[address.id];
  }

  /**
   * Get all objects stored in the map, in some unspecified order.
   */
  getAll(): T[] {
    const result = [];
    Object.keys(this._data).forEach((pluginName) => {
      const dataForPluginName = this._data[pluginName];
      Object.keys(dataForPluginName).forEach((type) => {
        const dataForType = dataForPluginName[type];
        Object.keys(dataForType).forEach((id) => {
          result.push(dataForType[id]);
        });
      });
    });
    return result;
  }

  /**
   * Remove any object with the given address. If none exists, this
   * method does nothing.
   */
  remove(address: Address): this {
    if (address == null) {
      throw new Error(`address is ${String(address)}`);
    }
    const dataForPluginName = this._data[address.pluginName];
    if (dataForPluginName === undefined) {
      return this;
    }
    const dataForType = dataForPluginName[address.type];
    if (dataForType === undefined) {
      return this;
    }
    delete dataForType[address.id];
    if (Object.keys(dataForType).length === 0) {
      delete dataForPluginName[address.type];
      if (Object.keys(dataForPluginName).length === 0) {
        delete this._data[address.pluginName];
      }
    }
    return this;
  }
}

export function toString(x: Address): string {
  return stringify(x);
}

export function fromString(s: string): Address {
  return JSON.parse(s);
}
