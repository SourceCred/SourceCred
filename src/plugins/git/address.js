// @flow

import type {Address} from "../../core/address";
import type {EdgeType, NodeType} from "./types";
import {COMMIT_NODE_TYPE, GIT_PLUGIN_NAME} from "./types";

export function _makeAddress(type: NodeType | EdgeType, id: string): Address {
  return {
    pluginName: GIT_PLUGIN_NAME,
    type,
    id,
  };
}

export function commitAddress(hash: string): Address {
  return _makeAddress(COMMIT_NODE_TYPE, hash);
}
