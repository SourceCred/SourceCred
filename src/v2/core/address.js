// @flow

import deepEqual from "lodash.isequal";
import stringify from "json-stable-stringify";

import {toCompat, fromCompat} from "../../v1/util/compat";
import type {Compatible} from "../../v1/util/compat";

export type PluginType = {|+plugin: string, +type: string|};

export type Address = {|
  +owner: PluginType,
  +id: string,
|};

export interface Addressable {
  +address: Address;
}

export type SansAddress<T: Addressable> = $Exact<$Diff<T, {+address: Address}>>;

export type AddressMapJSON<T: Addressable> = Compatible<{
  [serializedAddress: string]: SansAddress<T>,
}>;

export const COMPAT_TYPE = "sourcecred/sourcecred/AddressMap";
export const COMPAT_VERSION = "0.2.0";

/**
 * A data structure for storing addressable objects, keyed by their
 * addresses.
 */
export class AddressMap<T: Addressable> {
  /*
   * Nested structure for fast access.
   *
   * It is an representation invariant that there are no empty objects
   * in this structure (except possibly the top-level object). In
   * particular, if `_data[somePlugin]` is not `undefined`, then it
   * is a non-empty object. This is required so that two `AddressMap`s
   * are logically equal exactly if their `_data` fields are deep-equal.
   *
   * This nested structure is significantly more performant than a
   * simpler version using a flat object keyed by `stringify(address)`.
   * For basic performance tests, see:
   *     https://jsperf.com/address-string-302039074
   */
  _data: {[plugin: string]: {[type: string]: {[id: string]: T}}};

  /**
   * Create an empty `AddressMap`.
   */
  constructor() {
    this._data = {};
  }

  /**
   * Test whether this map logically equals another map. Two maps are
   * logically equal if they contain the same keys and the values at
   * each key are deep-equal.
   */
  equals(that: AddressMap<T>): boolean {
    return deepEqual(this._data, that._data);
  }

  toJSON(): AddressMapJSON<T> {
    const result = {};
    Object.keys(this._data).forEach((plugin) => {
      const dataForPlugin = this._data[plugin];
      Object.keys(dataForPlugin).forEach((type) => {
        const dataForType = dataForPlugin[type];
        Object.keys(dataForType).forEach((id) => {
          const owner = {plugin, type};
          const address = {owner, id};
          const datum = {...dataForType[id]};
          delete datum.address;
          result[toString(address)] = datum;
        });
      });
    });
    return toCompat(
      {
        type: COMPAT_TYPE,
        version: COMPAT_VERSION,
      },
      result
    );
  }

  static fromJSON(json: AddressMapJSON<T>): AddressMap<T> {
    const decompat = fromCompat(
      {
        type: COMPAT_TYPE,
        version: COMPAT_VERSION,
      },
      json
    );
    const result: AddressMap<T> = new AddressMap();
    Object.keys(decompat).forEach((key) => {
      result.add({...decompat[key], address: JSON.parse(key)});
    });
    return result;
  }

  /**
   * Add the given object to the map, replacing any existing value for
   * the same address.
   *
   * Returns `this` for easy chaining.
   */
  add(t: T): this {
    const {address} = t;
    if (address == null) {
      throw new Error(`address is ${String(t.address)}`);
    }
    let dataForPlugin = this._data[address.owner.plugin];
    if (dataForPlugin === undefined) {
      this._data[address.owner.plugin] = dataForPlugin = {};
    }
    let dataForType = dataForPlugin[address.owner.type];
    if (dataForType === undefined) {
      dataForPlugin[address.owner.type] = dataForType = {};
    }
    dataForType[address.id] = t;
    return this;
  }

  /**
   * Get the object at the given address, if it exists, or `undefined`
   * otherwise.
   */
  get(address: Address): T {
    if (address == null) {
      throw new Error(`address is ${String(address)}`);
    }
    const dataForPlugin = this._data[address.owner.plugin];
    if (dataForPlugin === undefined) {
      return (undefined: any);
    }
    const dataForType = dataForPlugin[address.owner.type];
    if (dataForType === undefined) {
      return (undefined: any);
    }
    return dataForType[address.id];
  }

  /**
   * Get all objects stored in the map, in some unspecified order.
   */
  getAll(): T[] {
    const result = [];
    Object.keys(this._data).forEach((plugin) => {
      const dataForPlugin = this._data[plugin];
      Object.keys(dataForPlugin).forEach((type) => {
        const dataForType = dataForPlugin[type];
        Object.keys(dataForType).forEach((id) => {
          result.push(dataForType[id]);
        });
      });
    });
    return result;
  }

  /**
   * Remove any object with the given address. If none exists, this
   * method does nothing.
   */
  remove(address: Address): this {
    if (address == null) {
      throw new Error(`address is ${String(address)}`);
    }
    const dataForPlugin = this._data[address.owner.plugin];
    if (dataForPlugin === undefined) {
      return this;
    }
    const dataForType = dataForPlugin[address.owner.type];
    if (dataForType === undefined) {
      return this;
    }
    delete dataForType[address.id];
    if (Object.keys(dataForType).length === 0) {
      delete dataForPlugin[address.owner.type];
      if (Object.keys(dataForPlugin).length === 0) {
        delete this._data[address.owner.plugin];
      }
    }
    return this;
  }
}

export function toString(x: Address): string {
  return stringify(x);
}

export function fromString(s: string): Address {
  return JSON.parse(s);
}
