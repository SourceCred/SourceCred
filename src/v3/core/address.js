// @flow
// This module implements Address functionality for the Graph module
// This module should not be directly imported by clients; rather, all public
// parts of this module are re-exported via Graph

import stringify from "json-stable-stringify";

export opaque type _NodeAddress = string;
export opaque type _EdgeAddress = string;
type GenericAddress = _NodeAddress | _EdgeAddress;

export const _Address = {
  toParts,
  nodeAddress,
  edgeAddress,
};

const NODE_PREFIX = "N";
const EDGE_PREFIX = "E";
const SEPARATOR = "\0";

const NODE = Symbol("NODE");
const EDGE = Symbol("EDGE");
type AddressType = typeof NODE | typeof EDGE;

function addressType(x: string): ?AddressType {
  if (x != null && x.endsWith(SEPARATOR)) {
    if (x.startsWith(NODE_PREFIX)) {
      return NODE;
    }
    if (x.startsWith(EDGE_PREFIX)) {
      return EDGE;
    }
  }
}

export function assertNodeAddress(x: _NodeAddress) {
  const type = addressType(x);
  switch (type) {
    case NODE:
      return;
    case EDGE:
      throw new Error(`Expected NodeAddress, got EdgeAddress: ${x}`);
    default:
      throw new Error(`Bad address: ${x}`);
  }
}

export function assertEdgeAddress(x: _EdgeAddress) {
  const type = addressType(x);
  switch (type) {
    case NODE:
      throw new Error(`Expected EdgeAddress, got NodeAddress: ${x}`);
    case EDGE:
      return;
    default:
      throw new Error(`Bad address: ${x}`);
  }
}
function assertAddress(x: GenericAddress) {
  if (addressType(x) == null) {
    throw new Error(`Expected Address, got: ${x}`);
  }
}

function assertAddressArray(arr: $ReadOnlyArray<string>) {
  if (arr == null) {
    throw new Error(String(arr));
  }
  arr.forEach((s: string) => {
    if (s == null) {
      throw new Error(`${String(s)} in ${stringify(arr)}`);
    }
    if (s.indexOf(SEPARATOR) !== -1) {
      throw new Error(`NUL char: ${stringify(arr)}`);
    }
  });
}

function nodeAddress(arr: $ReadOnlyArray<string> | void): _NodeAddress {
  if (arr === undefined) {
    return NODE_PREFIX + SEPARATOR;
  } else {
    assertAddressArray(arr);
    return [NODE_PREFIX, ...arr, ""].join(SEPARATOR);
  }
}

function edgeAddress(arr: $ReadOnlyArray<string> | void): _EdgeAddress {
  if (arr === undefined) {
    return EDGE_PREFIX + SEPARATOR;
  } else {
    assertAddressArray(arr);
    return [EDGE_PREFIX, ...arr, ""].join(SEPARATOR);
  }
}

function toParts(a: GenericAddress): string[] {
  assertAddress(a);
  const parts = a.split(SEPARATOR);
  return parts.slice(1, parts.length - 1);
}
