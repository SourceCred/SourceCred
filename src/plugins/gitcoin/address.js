// @flow

import {NodeAddress, type NodeAddressT} from "../../core/graph";

import {userNodeType, postNodeType, commentNodeType} from "./declaration";

import {type PostActivityId, type CommentId, type Username} from "./fetch";

export function userAddress(
  serverUrl: string,
  username: Username
): NodeAddressT {
  // TODO: potentially add userID
  return NodeAddress.append(
    userNodeType.prefix, // [sourceCred, gitcoin, user]
    serverUrl,
    String(username)
  );
}

export function postAddress(
  serverUrl: string,
  postId: PostActivityId
): NodeAddressT {
  return NodeAddress.append(
    postNodeType.prefix, // [sourceCred, gitcoin, post]
    serverUrl,
    String(postId)
  );
}

export function commentAddress(
  serverUrl: string,
  commentId: CommentId
): NodeAddressT {
  return NodeAddress.append(
    commentNodeType.prefix, // [sourceCred, gitcoin, comment]
    serverUrl,
    String(commentId)
  );
}
