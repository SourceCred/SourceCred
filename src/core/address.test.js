// @flow

import sortBy from "lodash.sortby";
import stringify from "json-stable-stringify";

import type {Address} from "./address";
import {AddressMap, fromString, toString} from "./address";

describe("address", () => {
  // Some test data using objects that have addresses, like houses.
  type House = {|
    +address: Address,
    +beds: number,
    +baths: number,
  |};
  function makeAddress(type: "HOME" | "BUSINESS", id: string): Address {
    return {
      repositoryName: "sourcecred/suburbia",
      pluginName: "houseville",
      id,
      type,
    };
  }
  const mansion = (): House => ({
    address: makeAddress("HOME", "mansion"),
    beds: 10,
    baths: 5,
  });
  const fakeMansion = (): House => ({
    // Same address, different content.
    address: makeAddress("HOME", "mansion"),
    beds: 33,
    baths: 88,
  });
  const mattressStore = (): House => ({
    address: makeAddress("BUSINESS", "mattressStore"),
    beds: 99,
    baths: 1,
  });

  describe("AddressMap", () => {
    const makeMap = (): AddressMap<House> =>
      new AddressMap().add(mansion()).add(mattressStore());

    it("creates a simple map", () => {
      makeMap();
    });

    it("gets objects by key", () => {
      expect(makeMap().get(mansion().address)).toEqual(mansion());
      expect(makeMap().get(mattressStore().address)).toEqual(mattressStore());
    });

    it("gets all objects, in some order", () => {
      const actual = makeMap().getAll();
      const expected = [mansion(), mattressStore()];
      const sort = (xs) => sortBy(xs, (x) => stringify(x.address));
      expect(sort(actual)).toEqual(sort(expected));
    });

    it("removes objects by key", () => {
      expect(
        makeMap()
          .remove(mansion().address)
          .get(mansion().address)
      ).toBeUndefined();
    });

    it("stringifies to JSON", () => {
      expect(makeMap().toJSON()).toMatchSnapshot();
    });

    it("stringifies elements sans addresses", () => {
      const json = makeMap().toJSON();
      Object.keys(json).forEach((k) => {
        const value = json[k];
        expect(Object.keys(value).sort()).toEqual(["baths", "beds"]);
      });
    });

    it("rehydrates elements with addresses", () => {
      const newMap: AddressMap<House> = AddressMap.fromJSON(makeMap().toJSON());
      newMap.getAll().forEach((house) => {
        expect(Object.keys(house).sort()).toEqual(["address", "baths", "beds"]);
      });
    });

    it("preserves equality over a JSON roundtrip", () => {
      const result = AddressMap.fromJSON(makeMap().toJSON());
      expect(result.equals(makeMap())).toBe(true);
    });

    it("recognizes reference equality", () => {
      const x = makeMap();
      expect(x.equals(x)).toBe(true);
    });

    it("recognizes deep equality", () => {
      expect(makeMap().equals(makeMap())).toBe(true);
    });

    it("recognizes equality invariant of construction order", () => {
      const m1 = new AddressMap().add(mansion()).add(mattressStore());
      const m2 = new AddressMap().add(mattressStore()).add(mansion());
      expect(m1.equals(m2)).toBe(true);
      expect(m2.equals(m1)).toBe(true);
    });

    it("recognizes disequality when element lists differ", () => {
      expect(makeMap().equals(new AddressMap())).toBe(false);
      expect(new AddressMap().equals(makeMap())).toBe(false);
    });

    it("recognizes disequality when contents differ", () => {
      const m1 = new AddressMap().add(mattressStore()).add(mansion());
      const m2 = new AddressMap().add(mattressStore()).add(fakeMansion());
      expect(m1.equals(m2)).toBe(false);
      expect(m2.equals(m1)).toBe(false);
    });

    describe("has nice error messages", () => {
      [null, undefined].forEach((bad) => {
        it(`when getting ${String(bad)} elements`, () => {
          const message = `address is ${String(bad)}`;
          expect(() => makeMap().get((bad: any))).toThrow(message);
        });
        it(`when removing ${String(bad)} elements`, () => {
          const message = `address is ${String(bad)}`;
          expect(() => makeMap().remove((bad: any))).toThrow(message);
        });
        it(`when adding elements with ${String(bad)} address`, () => {
          const message = `address is ${String(bad)}`;
          const element = {
            address: (bad: any),
            beds: 23,
            baths: 45,
          };
          expect(() => makeMap().add(element)).toThrow(message);
        });
      });
    });
  });

  describe("toString and fromString", () => {
    const examples = () => [mansion(), fakeMansion(), mattressStore()];
    it("simple round trips work", () => {
      examples().forEach((x) => {
        expect(x.address).toEqual(fromString(toString(x.address)));
      });
    });
    it("serialization looks good in snapshot review", () => {
      const serialized = examples().map((x) => [
        x.address,
        toString(x.address),
      ]);
      expect(serialized).toMatchSnapshot();
    });
    it("Order of insertion does not matter", () => {
      const a1 = {
        pluginName: "foo",
        type: "bar",
        id: "zoombat",
        repositoryName: "oregano",
      };
      const a2 = {
        id: "zoombat",
        type: "bar",
        repositoryName: "oregano",
        pluginName: "foo",
      };
      expect(toString(a1)).toEqual(toString(a2));
    });
  });
});
