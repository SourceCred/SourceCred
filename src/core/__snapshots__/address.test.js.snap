// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`address AddressMap stringifies to JSON 1`] = `
Object {
  "{\\"id\\":\\"mansion\\",\\"pluginName\\":\\"houseville\\",\\"type\\":\\"HOME\\"}": Object {
    "baths": 5,
    "beds": 10,
  },
  "{\\"id\\":\\"mattressStore\\",\\"pluginName\\":\\"houseville\\",\\"type\\":\\"BUSINESS\\"}": Object {
    "baths": 1,
    "beds": 99,
  },
}
`;

exports[`address toString and fromString serialization looks good in snapshot review 1`] = `
Array [
  Array [
    Object {
      "id": "mansion",
      "pluginName": "houseville",
      "type": "HOME",
    },
    "{\\"id\\":\\"mansion\\",\\"pluginName\\":\\"houseville\\",\\"type\\":\\"HOME\\"}",
  ],
  Array [
    Object {
      "id": "mansion",
      "pluginName": "houseville",
      "type": "HOME",
    },
    "{\\"id\\":\\"mansion\\",\\"pluginName\\":\\"houseville\\",\\"type\\":\\"HOME\\"}",
  ],
  Array [
    Object {
      "id": "mattressStore",
      "pluginName": "houseville",
      "type": "BUSINESS",
    },
    "{\\"id\\":\\"mattressStore\\",\\"pluginName\\":\\"houseville\\",\\"type\\":\\"BUSINESS\\"}",
  ],
]
`;
